tests/test__io_utils.py

```diff
@@ -30,6 +38,16 @@ def test_read_return_reader(uses_sample_plugin):
     assert reader.command == f"{SAMPLE_PLUGIN_NAME}.some_reader"
 
 
+def test_read_uppercase_extension_return_reader(uses_sample_plugin):
```

Comment:
Ok now that I look at it I think this is kind of a misleading way to check for case sensitivity. I think we should update the sample plugin to be case insensitive, or only pass it lowercase paths - it's not great to co-opt the exception to test case sensitivity. In this PR we should instead update [this test](https://github.com/napari/npe2/blob/14b2c055865001a60b7e15af90bd38771c75d41a/tests/test_contributions.py#L62) to check case insensitivity (or maybe just add a new test using that pattern).

This test also tries to make sure that we pass the full path, but it doesn't really achieve this, since it's `iter_compatible_readers` that **might** change it, and that's not called here. Testing for this should occur on the `napari` side. So I think here we should actually just remove this test

Reply:
I did it this way to ensure that the matching was insensitive, but that the real path was passed so a plugin could be case sensitive internally. I thought this was what we wanted from the community meeting 😅

Reply:
Oh wait, I got confused heh 😅 . So yes, we do definitely want to test that the path casing is preserved when passed to the read function.

I do still think this is a slightly odd way of testing it - i.e. it relies on a side effect of the sample plugin. So my recommendation for testing that the path is unchanged is that we use a dynamic plugin - I just tried doing that though and it didn't work so... please hold hehe

Reply:
Ok @psobolewskiPhD after #285 is merged, you should be able to write a test similar to this:

```ipython
In [2]: from npe2.io_utils import _read

In [3]: from npe2 import PluginManager, DynamicPlugin

In [4]: pm = PluginManager()

In [5]: plugin = DynamicPlugin('tif-read', plugin_manager=pm)

In [6]:     @plugin.contribute.reader
   ...:     def read_path(path):
   ...:         if path.lower() != path:
   ...:             raise ValueError("Path has capitals in it as we expected!")
   ...:         def _reader(path):
   ...:             return [(None,)]
   ...: 
   ...:         return _reader
   ...: 
 # we now expect this to raise a ValueError because the path has capitals in it, so you could use `with pytest.raises()` here
In [7]: _read(["some.BLAH"], stack=False, _pm=pm, plugin_name='tif-read')
```

Of course you don't **have** to raise an error, you could return a read function that returned some kind of known layer data that you then check against. But an error raised is easy to check for. Lmk if this makes sense!

Reply:
@DragaDoncila I'm not sure I totally get it. This creates a reader, but one that has no filename pattern right? So the changes implemented in this PR won't be tested I dont think and this test will always raise the error, so always pass?
Maybe I should edit the sample plugin to check the path that is passed (like it does) but not error and just report back that it was upper or lower?

Reply:
Sorry @psobolewskiPhD you can provide filename patterns in the reader contribution's decorator - I was messing around with other things and forgot to add the filename pattern back sorry. If you search the code base for `DynamicPlugin` you should see some examples. I can post the full test later, I'm just on my phone. This approach is also how we should test the other suggestions from Grzegorz.

Reply:
```ipython
In [1]: from npe2 import PluginManager, DynamicPlugin

In [2]: pm = PluginManager()

In [3]: reader_plugin = DynamicPlugin(name='tif_reader', plugin_manager=pm)

In [4]: @reader_plugin.contribute.reader(filename_patterns=["*.tif"], accepts_directories=False)
   ...: def get_read(pths):
   ...:     # we passed a path with capitals, and it should stil contain them
   ...:     if pths.lower() == pths:
   ...:         raise ValueError("Path was changed before being passed to reader")
   ...:     def read(paths):
   ...:         return [(None,)]
   ...:     return read
   ...: 
```

Reply:
OK, that makes more sense and it's what I sort of dug down into reading the code, so I guess I'm on the right track. But when I do put filename_patterns there i keep getting
`ValueError: No readers returned data for ['some.JPG']` and `ValueError: No readers returned data for ['some.jpg']` regardless of the pattern, so I must still be doing something wrong. I'll take another look tomorrow.

Reply:
Hmmm interesting. Let me know if you wanna pair on this, otherwise I can also take a look tomorrow!

Reply:
We can pair, could be fun.
Alternately, maybe just implementing this in the sample_plugin would be more robust anyways? Just need to modify the manifest to include the other permitted patterns and tweak the code to not error on the upper case extension as it does now. Could then probably parameterize just the paths to test.

Reply:
@psobolewskiPhD this test passes locally for me and seems to do everything we need. I like using `DynamicPlugin` because it's very flexible. We can use the same pattern of test to try directories, capital pattern and lowercase path, etc. If we update the `sample_plugin` it's a lot more annoying to rig up all the different options.

```python
def test_read_uppercase_extension():
    pm = PluginManager()
    plugin = DynamicPlugin("tif-plugin", plugin_manager=pm)
    path = "something.TIF"

    # reader should be compatible despite lowercase pattern
    @plugin.contribute.reader(filename_patterns=["*.tif"])    
    def get_read(path):
        if path.lower() != path:
            # if this error is raised we can be certain path is unchanged
            raise ValueError("Given path contains capitals.")
        def read(paths):
            return [(None,)]
        return read

    with pytest.raises(ValueError, match="Given path contains capitals."):
        io_utils._read([path], stack=False, _pm=pm)
```